numpy/lib/format.py

```diff
@@ -304,27 +305,32 @@ def _write_array_header(fp, d, version=None):
         header.append("'%s': %s, " % (key, repr(value)))
     header.append("}")
     header = "".join(header)
-    # Pad the header with spaces and a final newline such that the magic
-    # string, the header-length short and the header are aligned on a
-    # 16-byte boundary.  Hopefully, some system, possibly memory-mapping,
-    # can take advantage of our premature optimization.
-    current_header_len = MAGIC_LEN + 2 + len(header) + 1  # 1 for the newline
-    topad = 16 - (current_header_len % 16)
-    header = header + ' '*topad + '\n'
-    header = asbytes(_filter_header(header))
+    header = _filter_header(header)
+
+    hlen = len(header) + 1 # 1 for newline
+    padlen1 = ARRAY_ALIGN - ((MAGIC_LEN + 2 + hlen) % ARRAY_ALIGN)
+    padlen2 = ARRAY_ALIGN - ((MAGIC_LEN + 4 + hlen) % ARRAY_ALIGN)
 
-    hlen = len(header)
-    if hlen < 256*256 and version in (None, (1, 0)):
+    if hlen + padlen1 < 2**16 and version in (None, (1, 0)):
         version = (1, 0)
-        header_prefix = magic(1, 0) + struct.pack('<H', hlen)
-    elif hlen < 2**32 and version in (None, (2, 0)):
+        header_prefix = magic(1, 0) + struct.pack('<H', hlen + padlen1)
+        topad = padlen1
+    elif hlen + padlen2 < 2**32 and version in (None, (2, 0)):
         version = (2, 0)
-        header_prefix = magic(2, 0) + struct.pack('<I', hlen)
+        header_prefix = magic(2, 0) + struct.pack('<I', hlen + padlen2)
+        topad = padlen2
     else:
         msg = "Header length %s too big for version=%s"
         msg %= (hlen, version)
         raise ValueError(msg)
 
+    # Pad the header with spaces and a final newline such that the magic
+    # string, the header-length short and the header are aligned on a
+    # ARRAY_ALIGN byte boundary.  This supports memory mapping on systems
+    # like Linux where mmap() offset must be page-aligned.
+    header = header + ' '*topad + '\n'
+    header = asbytes(header)
```

Comment:
Since we're touching this, let's kill `asbytes` and use `b'...'` strings directly